vault/acl.go

```diff
@@ -47,41 +55,141 @@ func NewACL(policies []*Policy) (*ACL, error) {
 			// Check for an existing policy
 			raw, ok := tree.Get(pc.Prefix)
 			if !ok {
-				tree.Insert(pc.Prefix, pc.CapabilitiesBitmap)
+				var entry aclEntry
+				if pc.MFAMethods != nil && len(pc.MFAMethods) != 0 {
+					// We have a MFA so we leave the bitmap at 0 and use the map
+					entry.capabilitiesToMFAMap = convertCapBitmapToMFAMap(pc.CapabilitiesBitmap, pc.MFAMethods)
+				} else {
+					// We don't have MFA so we use the straight bitmap to be speedier
+					entry.capabilitiesBitmap = pc.CapabilitiesBitmap
+				}
+				tree.Insert(pc.Prefix, entry)
 				continue
 			}
-			existing := raw.(uint32)
+			existing := raw.(aclEntry)
 
 			switch {
-			case existing&DenyCapabilityInt > 0:
+			case existing.capabilitiesBitmap&DenyCapabilityInt > 0:
 				// If we are explicitly denied in the existing capability set,
-				// don't save anything else
+				// don't save anything else. Explicit denial entries do not set
+				// the MFA map, so checking the bitmap is sufficient.
 
 			case pc.CapabilitiesBitmap&DenyCapabilityInt > 0:
-				// If this new policy explicitly denies, only save the deny value
-				tree.Insert(pc.Prefix, DenyCapabilityInt)
+				// If this new policy explicitly denies, only save the deny
+				// value and leave the map at nil
+				tree.Insert(pc.Prefix, aclEntry{
+					capabilitiesBitmap: DenyCapabilityInt,
+				})
 
 			default:
 				// Insert the capabilities in this new policy into the existing
 				// value
-				tree.Insert(pc.Prefix, existing|pc.CapabilitiesBitmap)
+				tree.Insert(pc.Prefix, *(mergeACLEntryCapabilities(&existing, pc)))
 			}
 		}
 	}
 	return a, nil
 }
 
+func mergeACLEntryCapabilities(existing *aclEntry, pc *PathCapabilities) (ret *aclEntry) {
```

Comment:
What are the semantics of the merge? If one policy requires MFA and another doesn't, it seems like the merger should drop the MFA requirement given that permissions are a union of the greatest allowed privilege with the exception of deny at highest precedence.